webxr/openxr/mod.rs

```diff
@@ -155,41 +173,119 @@ struct OpenXrDevice {
+        context: &mut surfman::Context,
+        _context_id: surfman::ContextID,
+        new_front_buffer: Surface,
+    ) -> Result<(), surfman::Error> {
+        // At this point the front buffer's contents are already present in the underlying openxr texture.
+        // We only need to store the surface because the webxr crate's API assumes that Surface objects
+        // must be passed to the rendering method.
+        self.surface = Some(new_front_buffer);
+        // TODO(jdm): Destroy all cached EGL surfaces in destroy_all_surfaces instead.
+        if let Some(surface) = self.recycled_surface.take() {
+            device.destroy_surface(context, surface)?;
         }
+        Ok(())
+    }
+
+    fn create_sized_surface(
+        &mut self,
+        _device: &mut surfman::Device,
+        _context: &mut surfman::Context,
+        _size: euclid::default::Size2D<i32>,
+    ) -> Result<Surface, surfman::Error> {
+        // All OpenXR-based surfaces are created once during session initialization; we cannot create new ones.
+        // This is only used when resizing, however, and OpenXR-based systems don't resize.
+        Err(surfman::Error::UnsupportedOnThisPlatform)
+    }
+
+    fn destroy_all_surfaces(
+        &mut self,
+        _device: &mut surfman::Device,
+        _context: &mut surfman::Context,
+    ) -> Result<(), surfman::Error> {
+        Ok(())
     }
 }
 
 impl OpenXrDevice {
     fn new(
-        gl: Rc<dyn Gl>,
+        gl_thread: Box<dyn GlThread>,
+        provider_registration: Box<dyn SurfaceProviderRegistration>,
         instance: Instance,
         granted_features: Vec<String>,
+        id: SessionId,
     ) -> Result<OpenXrDevice, Error> {
-        let read_fbo = gl.gen_framebuffers(1)[0];
-        debug_assert_eq!(gl.get_error(), gl::NO_ERROR);
-
-        let write_fbo = gl.gen_framebuffers(1)[0];
-        debug_assert_eq!(gl.get_error(), gl::NO_ERROR);
+        let (device_tx, device_rx) = crossbeam_channel::unbounded();
+        let (provider_tx, provider_rx) = crossbeam_channel::unbounded();
+        let _ = gl_thread.execute(Box::new(move || {
+            // Get the current surfman device and extract it's D3D device. This will ensure
+            // that the OpenXR runtime's texture will be shareable with surfman's surfaces.
+            let (device, mut context) = unsafe {
+                SurfmanDevice::from_current_context().expect("Failed to create graphics context!")
+            };
+            device.destroy_context(&mut context).unwrap();
+            let d3d_device = device.d3d11_device();
+            // Smuggle the pointer out as a usize value; D3D11 devices are threadsafe
+            // so it's safe to use it from another thread.
+            let _ = device_tx.send(d3d_device.as_raw() as usize);
+            let _ = provider_rx.recv();
+        }));
+        // Get the D3D11 device pointer from the webgl thread.
+        let device = device_rx.recv().unwrap();
```

Comment:
The fundamental point of these changes is to ensure that OpenXR uses the WebGL thread's d3d device, rather than an independent one. Without that sharing, the WebGL thread can't share the OpenXR textures.

Reply:
Can we do it the other way round, where WebGL uses the device provided by OpenXR?